---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# smaR: Simple Metagenomic Analysis in R

<!-- badges: start -->
<!-- badges: end -->

## Overview

smaR (Simple Metagenomic Analysis in R) is an R package designed to streamline and simplify metagenomic data analysis workflows.
It provides a set of tools for processing, analyzing, and visualizing metagenomic sequencing data, making it accessible to
researchers with varying levels of bioinformatics expertise.
It starts from already calculated and merged taxonomic profiles, i.e. a table with samples as rows and species/OTUs/ASVs as columns,
or viceversa, and numeric values that represent relative abundances. smaR contains functions to easily check the correctness of
the matrix, calculate alpha- and beta-diversity and, if a valid relative metadata table is provided, can perform Non-metric
Multidimensional Scaling (NMDS) and differential abundance analysis (DAA) with different methods, as well as plot their results.
In particular, it can calculate and represent the consensus between the results obtained by Maaslin2 and MaAsLin3.
The package aims at simplifying and homogenizing common metagenomic workflows.

## Installation

smaR is an R package, you can install the development version
from GitHub running in R or Rstudio:

``` r
devtools::install_github("NiCarlino/smaR")
```

### Requirements

smaR requires an R version > 4.4. See the DESCRIPTION file for the full list of other packages dependencies.

## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(smaR)
abundance_table <- import_abundance_table("path_to_your_table.tsv", rowtype = "samples")
```
It uploads an abundance table and check that the table is consistent with relative abundance data (e.g. numeric values, normalize, fill NAs)
and set the correct orientation for the following steps.

In order to calculate the Shannon Index for all the samples included in the abundance table, run:

```{r}
df_alpha <- calculate_alpha_diversity(abundance_table, metric ="shannon")
```

You can also specify other metrics ("richness" or "simpson"), or even more than one if you want
to compare the results of different approaches.

Similarly, to calculate the distances among samples you can run

```{r}
df_beta <- calculate_beta_diversity(abundance_table, metric ="shannon")
```

## Running a demo


## License

This project is licensed under the GPL-3 License - see the LICENSE file for details.

## Contact

NiccolÃ² Carlino
URL: https://github.com/NiCarlino/smaR
